Submit runtime compilation telemetry to statsd

ISauve opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succint form, consider
  opening multiple pull requests instead of a single one.
-->

Submits runtime compilation & kernel header downloading telemetry to `statsd`, using the `datadog.system_probe` prefix so customers don't get billed.

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

Allows customers (or us, for support tickets) to debug any runtime compilation or kernel header downloading errors. Our current telemetry tells us if there's failures and which org they're coming from, but not which specific host failed.

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

ISauve committed:
Submit runtime compilation telemetry to statsd

brycekahle commented:
I think we need to update the system-probe configuration of statsd to handle if `STATSD_URL` is set. I see that being used first in multiple places, before the individual config values for host and port.

brycekahle commented on pkg/network/tracer/tracer.go:
```diff
@@ -451,6 +453,18 @@ func (t *Tracer) getRuntimeCompilationTelemetry() map[string]network.RuntimeComp
 		}
 		result[assetName] = tm
+
+		statsdMetricPrefix := "datadog.system_probe." + assetName
```
I think we should have the asset name be a tag, rather than part of the metric name. WDYT?

ISauve replied:
Works for me!

brycekahle commented on pkg/network/tracer/tracer.go:
```diff
@@ -451,6 +453,18 @@ func (t *Tracer) getRuntimeCompilationTelemetry() map[string]network.RuntimeComp
 		result[assetName] = tm
+
+		statsdMetricPrefix := "datadog.system_probe." + assetName
+		tags := []string{fmt.Sprintf("version:%s", version.AgentVersion)}
```
```suggestion
		tags := []string{fmt.Sprintf("agent_version:%s", version.AgentVersion)}
```

brycekahle commented on pkg/network/tracer/tracer.go:
```diff
@@ -451,6 +453,18 @@ func (t *Tracer) getRuntimeCompilationTelemetry() map[string]network.RuntimeComp
+		tags := []string{fmt.Sprintf("version:%s", version.AgentVersion)}
+
+		if tm.RuntimeCompilationEnabled {
+			if err := statsd.Client.Gauge(statsdMetricPrefix + ".runtime_compilation_result", float64(tm.RuntimeCompilationResult), tags, 1); err != nil {
```
Does `Count` make more sense, since we tracking the result of an event?

ISauve replied:
If we use `Count`, the agent will add up all the values received in a time interval. Since we're reporting something which has a continuous value, `Gauge` is a better representation.

ISauve committed:
Update pkg/network/tracer/tracer.go

Co-authored-by: Bryce Kahle <bryce.kahle@datadoghq.com>

ISauve committed:
Include asset name as tag rather than in metric name

ISauve committed:
Submit result/reason as tags + submit metrics as count

ISauve committed:
Include platform tag

ISauve committed:
Update agent-payload dependency
Rename headersNotFound -> headersNotFoundDownloadDisabled
Add comment to exported value

brycekahle commented on pkg/ebpf/bytecode/runtime/runtime_asset.go:
```diff
@@ -118,3 +129,62 @@ func (a *RuntimeAsset) GetTelemetry() map[string]int64 {
+func (a *RuntimeAsset) SubmitTelemetry() {
+	tm := a.runtimeCompiler.GetRCTelemetry()
+
+	if tm.compilationEnabled {
```
nit: invert this with an early return
```
if !tm.compilationEnabled {
    return
}
```

ISauve committed:
Early return from SubmitTelemetry
